triplea-game/triplea
# Decouple lobby integration tests from file system

## Overview

There are currently two copies of _lobby.properties_:

1. _lobby/lobby.properties_: Used by the production/staging lobby.
1. _game-core/config/lobby/lobby.properties_: Used by the lobby integration tests and the Eclipse `LobbyRunner` launch configuration.

The motivation behind this PR is to get rid of copy (2) to avoid the confusion between the two files and to decouple the lobby integration tests from the file system so they aren't dependent on the content of _lobby.properties_.

This PR is primarily an exercise in dependency injection.  The core change is getting rid of the `LobbyContext` utility and propagating the `LobbyPropertyReader` dependency throughout the system.  This then allows the integration tests to specify an in-memory instance of `LobbyPropertyReader` in lieu of creating an instance from _lobby.properties_.

## Functional Changes

None.

## Refactoring Changes

1. Removed `LobbyContext` and replaced its usage by adding a dependency on `LobbyPropertyReader`.
    * The `LobbyPropertyReader` dependency is injected using constructor injection, where necessary.
    * In production code, `LobbyRunner` is now the only entity responsible for reading _lobby.properties_ and injecting the resulting `LobbyPropertyReader` into the root `LobbyServer` object.
    * In test code, the lobby integration tests now use an in-memory configuration available from `TestLobbyPropertyReaders#INTEGRATION_TEST`.
1. Removed the now unused _game-core/config/lobby/lobby.properties_ file.
1. Moved _lobby/lobby.properties_ to _lobby/config/lobby/lobby.properties_ so the same relative path will be used from both an IDE and the build artifact.
    * Updated Eclipse `LobbyRunner` launch configuration, as needed.
    * Updated Gradle build artifact, as needed.
1. Converted `Database` from a collection of utility methods to a class that encapsulates the `LobbyPropertyReader` dependency required to create a database connection.
1. Extracted a new superclass (`AbstractController`) for DAO implementations in order to localize the dependency on the new `Database` instance.

## Manual Testing Performed

I ran a local lobby using both the Eclipse launch configuration and the Gradle build artifact.  In both cases, the lobby properties file was found correctly, and the lobby started successfully.  I did some additional smoke testing to ensure a client could connect to the lobby and that the lobby database was exercised.
